mautic/mautic
# Messenger improvements

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [❌]
| New feature/enhancement? (use the a.x branch)      | [✅]
| Deprecations?                          | [❌]
| BC breaks? (use the c.x branch)        | [❌]
| Automated tests included?              | [✅] 
| Related user documentation PR URL      |  
| Related developer documentation PR URL |  
| Issue(s) addressed                     | 



#### Description:



This PR  **fixes** and **improves** several issues connected with the Symfony Messenger implementation within Mautic.

### Fixes:

1. There was no way to change the failure transport (One of the Doctrine options was hardcoded). UI configuration was introduced.
2. It was impossible to change the retry strategy options until the Symfony cache was rebuilt.

### Improvements

1. **Any messenger transport can be used/configured without the need of creating an extra plugin for each of them.**
    - I tested the Doctrine, Redis and Beanstalkd transports in my local environment.
    - Interestingly Redis transport does not even require running `composer install symfony/redis-messenger`. 
3. Queuing is disabled by default.
5. Test buttons for queuing configuration were added.
6. The codebase was simplified a lot.
7. Messenger and email DSNs can be overridden within tests.

#### Steps to test this PR:


1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Make sure the queuing for emails is disabled by default.
3. Configure one of the supported messenger transports (such as Doctrine, Redis, Beanstalkd, ...) for the email queue, start a consumer using `bin/console messenger:consume email` and check that emails are delivered via that queue.
4. Make sure that test buttons in the "Queue Settings" are working as expected. 
5. Configure one of the supported transports for the failure queue and check that messages that fail end up in that queue (you can misconfigure the email DSN to make the messages fail).
6. Try changing the "Retry strategy" settings and make sure these changes are applied without a need of clearing the Symfony cache.
